core/common/src/main/java/alluxio/wire/TieredIdentity.java

```diff
@@ -58,9 +64,37 @@ public TieredIdentity(List<LocalityTier> tiers) {
    * @return the corresponding wire type tiered identity
    */
   public static TieredIdentity fromThrift(alluxio.thrift.TieredIdentity tieredIdentity) {
+    if (tieredIdentity == null) {
+      return null;
+    }
     return new TieredIdentity(tieredIdentity.getTiers().stream()
-        .map(LocalityTier::fromThrift).collect(Collectors.toList())
-    );
+        .map(LocalityTier::fromThrift).collect(Collectors.toList()));
+  }
+
+  /**
+   * @param identities the tiered identities to compare to
+   * @return the identity closest to this one
+   */
+  public Optional<TieredIdentity> nearest(List<TieredIdentity> identities) {
+    Preconditions.checkState(!identities.isEmpty(), "No identities given");
+    for (LocalityTier tier : mTiers) {
+      for (TieredIdentity identity : identities) {
+        for (LocalityTier otherTier : identity.mTiers) {
+          if (tier.mTierName.equals(otherTier.mTierName)
+              && tier.mValue != null
+              && tier.mValue.equals(otherTier.mValue)) {
+            return Optional.of(identity);
+          }
+        }
+      }
```

Comment:
Stick to using `stream` ?

Reply:
I don't see a way - this is quite complex